certora/ArrangerConduit.spec

```diff
@@ -0,0 +1,710 @@
+    function getFundRequestsLength() external returns (uint256) envfree;
+    function getFundRequest(uint256) external returns (IArrangerConduit.FundRequest) envfree;
+ 
+    // needed for resolving calls in the spec
+    function gem.balanceOf(address) external returns (uint256) envfree;
+    function gem.allowance(address, address) external returns (uint256) envfree;  
+    function registry.buffers(bytes32) external returns (address) envfree;
+    function roles.canCall(bytes32, address, address, bytes4) external returns (bool) envfree;
+    function aux.hashString(string) external returns (bytes32) envfree;
+
+    // needed for resolving calls in the contract
+    function _.balanceOf(address) external => DISPATCHER(true) UNRESOLVED;
+    function _.transfer(address, uint256) external => DISPATCHER(true) UNRESOLVED;
+    function _.transferFrom(address, address, uint256) external => DISPATCHER(true) UNRESOLVED;    
+}
+
+definition min(mathint x, mathint y) returns mathint = x < y ? x : y;
+
+// -----------------------------------------------------------------------------
+/**
+ * Verifying the string encoding
+ * =============================
+ * Whenever Solidity reads a string from storage, it verifies that it is properly
+ * encoded, otherwise it reverts. Note that this occurs also when writing a string
+ * to the storage.
+ * The encoding of a string is as follows:
+ * 1. If the first 32-bytes are even (the very last bit is 0) then:
+ *    - The length of the string is less than 32 bytes
+ *    - The first 31 bytes contain the string
+ *    - The last byte is twice the length of the string
+ * 2. If the first 32-bytes are odd (the very last bit is 1) then:
+ *    - The length of the string is 32 bytes or more
+ *    - The first 32-bytes are twice the length of the string plus 1!
+ *    - The string itself is in the following bytes
+ *
+ * So, an encoding is illegal if either:
+ * - the first 32-bytes are even and the value of the last byte >= 64, or
+ * - the first 32-bytes are odd but as an integer they are less than 65.
+ *
+ * One solution is to require that there are no illegal strings in storage, using
+ * a hook.
+ */
+
+
+/** This hook is called whenever the `info` field of a `FundRequest` in `fundRequests`
+ *  is read. To be precise, whenever the first 32 bytes of `info` are read.
+ *  To create the hook we needed to find the offset of `info` in the `FundRequest`
+ *  struct, here is the calculation:
+    struct FundRequest {
+        StatusEnum status; // 1 byte
+        address    asset;  // 20 bytes
+        bytes32    ilk;  // 32 bytes
+        uint256    amountRequested; // 32 bytes
+        uint256    amountFilled;  // 32 bytes
+        string     info;
+    }
+  * The storage is 32-bytes aligned. If each field gets 32-bytes, then the offset
+  * would be 32 * 5 = 160. If the `status` and `asset` fields are packed together,
+  * then the offset would be 32 * 4 = 128.
+  * Trial and error shows the offset is 128.
```

Comment:
Not sure we really need to leave this as a comment as we know that status and asset will be packed in one storage slot.

Reply:
Done